actionpack/lib/action_controller/metal/renderers.rb

```diff
@@ -103,17 +123,88 @@ def self.remove(key)
+    # rather than requiring one to define a method +_render_with_renderer_json+
+    # in a subclass and optionally call super on it.
+    #
+    # A principal benefit of this approach is that it promotes serialization on an object
+    # to a clearly-defined public interface, rather than requiring one to understand that
+    # calling, e.g. +render json: object+ calls +_render_to_body_with_renderer(options)+
+    # which calls +_render_with_renderer_#{key}+ where key is +json+, which is the method
+    # defined by calling +ActionController::Renderers.add :json+.
+    #
+    # ActiveModel::Serializers, for example, which has relied upon defining
+    # +_render_option_json+ and +_render_with_renderer_json+ in the controller, and calling +super+
+    # on the serialized object, now only needs to call
+    # +ActionController::Serializers.remove_serializer :json+ and define a new serializer with
+    # +ActionController::Serializers.add_serializer json do |json, options|+.
+    #
+    # See https://groups.google.com/forum/#!topic/rubyonrails-core/K8t4-DZ_DkQ/discussion for
+    # more background information.
+    module Serializers
+      extend ActiveSupport::Concern
+
+      included do
+        class_attribute :_serializers
+        self._serializers = Set.new.freeze
+      end
+
+      module ClassMethods
+        def use_serializers(*args)
+          serializers = _serializers + args
+          self._serializers = serializers.freeze
+        end
+        alias use_serializer use_serializers
+      end
+
+      # A Set containing serializer names that correspond to available serializer procs.
+      # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
+      SERIALIZERS = Set.new
+
+      def self._serialize_with_serializer_method_name(key)
+        "_serialize_with_serializer_#{key}"
+      end
+
+      def self.add_serializer(key, &block)
+        define_method(_serialize_with_serializer_method_name(key), &block)
+        SERIALIZERS << key.to_sym
+      end
+
+      # This method is the opposite of add_serializer method.
+      #
+      # To remove a csv serializer:
+      #
+      #   ActionController::Serializers.remove_serializer(:csv)
+      def self.remove_serializer(key)
+        SERIALIZERS.delete(key.to_sym)
+        method_name = _serialize_with_serializer_method_name(key)
+        remove_method(method_name) if method_defined?(method_name)
+      end
+    end
+
+
     module All
       extend ActiveSupport::Concern
       include Renderers
+      include Renderers::Serializers
 
       included do
         self._renderers = RENDERERS
+        self._serializers = SERIALIZERS
       end
     end
 
+    add_serializer :json do |json, options|
+      json.kind_of?(String) ? json : json.to_json(options)
+    end
+
     add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+      json = _serialize_with_serializer_json(json, options)
```

Comment:
Although I'm not a *huge* fan of this interface, I think it has the following benefits

1. It serves to implicitly document the use of methods like `_serialize_with_serializer_json`, which makes them easier to search for
1. It's a very small change
1. I think it could be a good thing to separate the ides of a 'json renderer' that uses as 'json serializer', yet keep them tightly coupled in usage and code location, even though there is a slightly higher cognitive burden.  Perhaps we should ensure the renderer has a helpful failure message if the serializer isn't defined.